### ISSUE TITLE
Setting up a different IntrospectionEndpoint ... ?

### ISSUE THREAD

**Issue by mfriese** [IdentityServer/IdentityServer4#4270]
My scenario is this: On a server there is an API project, an IdentityServer4 and an Angular Client running. Also there is a load balancer that does URL rewriting.

The IdentityServer4 can be reached from "outside" (web) by the URL "https://10.234.135.236:8098".
The URL will be rewritten "inside" the machine to "http://10.233.177.20:8098".

For that reason I've added IdentityServer config:
```
[...]
var builder = services.AddIdentityServer(options =>
{
    options.IssuerUri = "https://10.234.135.236:8098";
    options.PublicOrigin = "https://10.234.135.236:8098";
})
[...]
```
So when coming from the Angular Client application (web) and opening the IdentityServer4 Website (Quickstart MVC) logging in works fine. I am redirected to my angular client application and I am logged in with a proper bearer token.

What does no work is validating the token through the back-channel.
```
services.AddAuthentication(IdentityServerAuthenticationDefaults.AuthenticationScheme)
                .AddIdentityServerAuthentication(options =>
{
    options.ApiName = ...;
    options.Authority = "http://10.233.177.20:8098";
    options.RequireHttpsMetadata = ...;
    options.EnableCaching = ...;
    options.CacheDuration = TimeSpan.FromMinutes(settings.Value.AuthorizationCacheDuration);
});
```
The API project will call
```
http://10.233.177.20:8098/.well-known/openid-configuration
```
and this returns:
```
issuer | "https://10.234.135.236:8098"
jwks_uri | "https://10.234.135.236:8098/.well-known/openid-configuration/jwks"
authorization_endpoint | "https://10.234.135.236:8098/connect/authorize"
token_endpoint | "https://10.234.135.236:8098/connect/token"
userinfo_endpoint | "https://10.234.135.236:8098/connect/userinfo"
end_session_endpoint | "https://10.234.135.236:8098/connect/endsession"
check_session_iframe | "https://10.234.135.236:8098/connect/checksession"
revocation_endpoint | "https://10.234.135.236:8098/connect/revocation"
introspection_endpoint | "https://10.234.135.236:8098/connect/introspect"
device_authorization_endpoint | "https://10.234.135.236:8098/connect/deviceauthorization"
[...]
```
So when the API project tries to validate the token through the back-channel it tries to contract
```
https://10.234.135.236:8098
```
which is unreachable inside the server.

So I came across this issue: https://github.com/IdentityServer/IdentityServer4/issues/1627 and I was wondering if it is possible to set **IntrospectionEndpoint** directly. If not, how can I configure a different endpoint for the back-channel token validation? I've tried to figure this out for quite a while.

Thanks in advance!

**Comment by mfriese** [comment]
As a test I implemented a custom generator for the discovery document but I expected this to be configurable. So this is a working solution now, but is there no other way?

The addresses have to be removed of course.

```
namespace AuthorizationServer
{
    public class DiscoveryResponseGen : DiscoveryResponseGenerator
    {
        public DiscoveryResponseGen(IdentityServerOptions options,
            IResourceStore resourceStore,
            IKeyMaterialService keys,
            ExtensionGrantValidator extensionGrants,
            SecretParser secretParsers,
            IResourceOwnerPasswordValidator resourceOwnerValidator,
            ILogger<DiscoveryResponseGenerator> logger) :
            base(
                options,
                resourceStore,
                keys,
                extensionGrants,
                secretParsers,
                resourceOwnerValidator,
                logger)
        {
        }

        public override async Task<Dictionary<string, object>> CreateDiscoveryDocumentAsync(string baseUrl, string issuerUri)
            => await base.CreateDiscoveryDocumentAsync("http://10.233.177.20:8098/", "https://10.234.135.236:8098");
    }
}
```

**Comment by leastprivilege** [comment]
The OIDC protocol demands the use of a stable "authority" - typically the DNS name of the server using "https://". I wouldn't recommend using IP addresses.

If you need to run on 2 physical IPs - have a DNS entry that resolves to the corresponding IPs for each network.

**Comment by mfriese** [comment]
Thank you! So I will have a server name at some point. It's just that IdentityServer and the Backend run on the same Machine. The frontend is available on the internet.

from the backend I have to use ```http://internal_ip:port/``` The frontend on the other hand uses ```https://external_address```. Now the problem arises, when the backend tries to contact ```https://external_address```, which ist unreachable. But the frontend has to use ```https://external_address``` for logging in. I can't just open IS4 Website, cause it redirects to a wrong URL for the login page (which is ```http://internal_ip:port/``` that is not reachable from the web). So my only solution was to provide different endpoints for the website and the OIDC endpoints.

There will be domain names and https in the future. please ignore that part.

Basically like this:

```
public override async Task<Dictionary<string, object>> CreateDiscoveryDocumentAsync(string baseUrl, string issuerUri)
            => await base.CreateDiscoveryDocumentAsync("http://internal_ip:port/", "https://external_address");
```

**Comment by leastprivilege** [comment]
Yea - as I said - two DNS servers (internal and external) are the right solution.

**Comment by mfriese** [comment]
So basically I would set up for example ```https://mydomain.com``` and redirect it to identity server from inside and outside? That is the solution? Thank you very much!
